File: apps/HealthApp/android/app/src/main/java/com/myhealth/app/ui/profile/ProfileScreen.kt
package com.myhealth.app.ui.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.ViewModel
import com.myhealth.app.data.secure.InsuranceCardDao
import com.myhealth.app.data.secure.MyChartIssuerDao
import com.myhealth.app.data.secure.SecureTokenStore
import com.myhealth.app.fhir.EpicSandboxConfig
import com.myhealth.app.ui.theme.CarePlusColor
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.map

/**
 * Care+ Profile. Replaces the previous stub. Shows BMI auto-calc + a
 * profile-completion percentage so users know what's still missing
 * before week 2's MyChart-merge backfills more data.
 */
@Composable
fun ProfileScreen(vm: ProfileViewModel = hiltViewModel()) {
    var name by remember { mutableStateOf("") }
    var heightCm by remember { mutableFloatStateOf(170f) }
    var weightKg by remember { mutableFloatStateOf(65f) }
    val tint = CarePlusColor.CareBlue
    val sources by vm.sources.collectAsState(initial = emptyList())

    val bmi = if (heightCm > 0) weightKg / ((heightCm / 100f) * (heightCm / 100f)) else 0f
    val completion = listOf(name.isNotEmpty(), heightCm > 0, weightKg > 0)
        .count { it } * 100 / 3

    Column(Modifier.fillMaxSize().padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text("Profile", fontSize = 28.sp, fontWeight = FontWeight.Bold)

        Card(Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp)) {
            Column(Modifier.padding(16.dp)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text("Profile completion", modifier = Modifier.weight(1f))
                    Text("$completion%", color = tint, fontWeight = FontWeight.Bold)
                }
                Spacer(Modifier.height(6.dp))
                LinearProgressIndicator(
                    progress = { completion / 100f },
                    modifier = Modifier.fillMaxWidth(),
                    color = tint,
                )
                Text("Complete your profile so MyChart imports merge cleanly.",
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    fontSize = 11.sp, modifier = Modifier.padding(top = 6.dp))
            }
        }

        OutlinedTextField(value = name, onValueChange = { name = it },
            label = { Text("Name") }, modifier = Modifier.fillMaxWidth())

        Text("Height: ${heightCm.toInt()} cm", fontWeight = FontWeight.SemiBold)
        Slider(value = heightCm, onValueChange = { heightCm = it }, valueRange = 120f..220f)

        Text("Weight: ${weightKg.toInt()} kg", fontWeight = FontWeight.SemiBold)
        Slider(value = weightKg, onValueChange = { weightKg = it }, valueRange = 30f..200f)

        HorizontalDivider()

        Card(Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp)) {
            Row(Modifier.padding(16.dp).fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically) {
                Text("BMI", modifier = Modifier.weight(1f))
                Text("%.1f".format(bmi), fontWeight = FontWeight.Bold)
                Text(bmiLabel(bmi), color = bmiColor(bmi), fontWeight = FontWeight.SemiBold,
                    modifier = Modifier
                        .padding(start = 8.dp)
                        .background(bmiColor(bmi).copy(alpha = 0.18f), RoundedCornerShape(99.dp))
                        .padding(horizontal = 8.dp, vertical = 2.dp))
            }
        }

        // ── Connected sources ───────────────────────────────────
        Text("Connected sources", fontWeight = FontWeight.SemiBold,
            modifier = Modifier.padding(top = 8.dp))
        sources.forEach { src ->
            Card(Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp)) {
                Row(Modifier.padding(12.dp),
                    verticalAlignment = Alignment.CenterVertically) {
                    Text(src.label, modifier = Modifier.weight(1f))
                    when (src.status) {
                        SourceStatus.CONNECTED -> Text("✓",
                            color = CarePlusColor.Success, fontWeight = FontWeight.Bold)
                        SourceStatus.NOT_CONNECTED -> Text("Connect",
                            color = tint, fontWeight = FontWeight.SemiBold, fontSize = 12.sp)
                        SourceStatus.ADD -> Text("Add",
                            color = tint, fontWeight = FontWeight.SemiBold, fontSize = 12.sp)
                    }
                }
            }
        }
    }
}

data class ConnectedSource(val label: String, val status: SourceStatus)

enum class SourceStatus { CONNECTED, NOT_CONNECTED, ADD }

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val secureTokens: SecureTokenStore,
    private val insuranceDao: InsuranceCardDao,
    private val mychartDao: MyChartIssuerDao,
) : ViewModel() {
    val sources = mychartDao.observeAll().map { issuers ->
        listOf(
            ConnectedSource("Health Connect", SourceStatus.CONNECTED), // optimistic
            ConnectedSource("Epic MyChart",
                if (issuers.isNotEmpty()
                    || secureTokens.fhirAccessToken(EpicSandboxConfig.ISSUER) != null)
                    SourceStatus.CONNECTED else SourceStatus.NOT_CONNECTED),
            ConnectedSource("Insurance card",
                // Snapshot read; for week-1 we re-emit on each subscribe.
                if (kotlinx.coroutines.runBlocking { insuranceDao.latest() } != null)
                    SourceStatus.CONNECTED else SourceStatus.NOT_CONNECTED),
            ConnectedSource("Pharmacy", SourceStatus.ADD),
        )
    }
}

private fun bmiLabel(b: Float) = when {
    b < 18.5f -> "Under"
    b < 25f -> "Normal"
    b < 30f -> "Over"
    else -> "Obese"
}

private fun bmiColor(b: Float): Color = when {
    b < 18.5f -> CarePlusColor.Info
    b < 25f -> CarePlusColor.Success
    b < 30f -> CarePlusColor.Warning
    else -> CarePlusColor.Danger
}

File: apps/HealthApp/ios/FitFusion/Views/Profile/ProfileScreen.swift
import SwiftUI
import FitFusionCore

/// Care+ Profile screen reachable from the global header avatar on every
/// tab. Replaces the old `Views/More/HealthProfileView.swift` as the
/// primary "me" surface (HealthProfileView remains accessible inside this
/// screen as the dedicated medical conditions list).
///
/// Shows BMI auto-calc + a profile-completion percentage so users know
/// what's still missing before week 2's MyChart-merge backfills more data.
struct ProfileScreen: View {

    @StateObject private var hcStore = HealthConditionsStore.shared
    @Environment(\.dismiss) private var dismiss

    @AppStorage("profile.name") private var name: String = ""
    @AppStorage("profile.heightCm") private var heightCm: Double = 170
    @AppStorage("profile.weightKg") private var weightKg: Double = 65
    @AppStorage("profile.unitsImperial") private var unitsImperial: Bool = false

    private var bmi: Double {
        let h = heightCm / 100.0
        return h > 0 ? weightKg / (h * h) : 0
    }

    /// Crude completion estimate: 5 fields × 20% each.
    private var completionPercent: Int {
        var done = 0
        if !name.isEmpty { done += 1 }
        if heightCm > 0 { done += 1 }
        if weightKg > 0 { done += 1 }
        if hcStore.hasAnyCondition || hcStore.conditions == [.none] { done += 1 }
        if hcStore.lastDoctorReview != nil { done += 1 }
        return Int(Double(done) / 5.0 * 100.0)
    }

    var body: some View {
        Form {
            Section {
                completionCard
                    .listRowInsets(EdgeInsets())
                    .listRowBackground(Color.clear)
            }

            Section("Personal") {
                TextField("Name", text: $name)
                Stepper(value: $heightCm, in: 120...220, step: 1) {
                    Text("Height: \(Int(heightCm)) cm")
                }
                Stepper(value: $weightKg, in: 30...200, step: 0.5) {
                    Text(String(format: "Weight: %.1f kg", weightKg))
                }
                Toggle("Imperial units (ft / lb)", isOn: $unitsImperial)
            }

            Section("Connected sources") {
                connectedRow(symbol: "heart.text.square.fill",
                             label: "Apple Health",
                             status: .connected)
                connectedRow(symbol: "cross.case.fill",
                             label: "Epic MyChart",
                             status: KeychainStore.shared.fhirAccessToken(
                                issuer: EpicSandboxConfig.issuer
                             ) != nil ? .connected : .notConnected)
                connectedRow(symbol: "creditcard.fill",
                             label: "Insurance card",
                             status: PHIStore.shared.latestInsuranceCard() != nil
                                ? .connected : .notConnected)
                connectedRow(symbol: "pills.fill",
                             label: "Pharmacy",
                             status: .add)
            }

            Section("Body composition") {
                HStack {
                    Text("BMI")
                    Spacer()
                    Text(String(format: "%.1f", bmi)).foregroundStyle(.secondary)
                    Text(bmiLabel(bmi))
                        .font(.caption.weight(.semibold))
                        .padding(.horizontal, 8).padding(.vertical, 2)
                        .background(bmiColor(bmi).opacity(0.18), in: Capsule())
                        .foregroundStyle(bmiColor(bmi))
                }
            }

            Section("Health profile") {
                NavigationLink("Conditions & medical history") { HealthProfileView() }
            }

            Section("Account") {
                NavigationLink("Settings") { SettingsView() }
            }

            Section {
                Button(role: .destructive) {
                    // Auth / wipe handled via environment elsewhere; this is
                    // just the placeholder hook so the screen renders the row.
                } label: {
                    Label("Sign out", systemImage: "rectangle.portrait.and.arrow.right")
                }
            }
        }
        .navigationTitle("Profile")
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Button("Done") { dismiss() }
            }
        }
    }

    private var completionCard: some View {
        VStack(alignment: .leading, spacing: CarePlusSpacing.sm) {
            HStack {
                Text("Profile completion").font(.subheadline)
                Spacer()
                Text("\(completionPercent)%").font(.headline)
                    .foregroundStyle(CarePlusPalette.careBlue)
            }
            ProgressView(value: Double(completionPercent), total: 100)
                .tint(CarePlusPalette.careBlue)
            Text("Complete your profile so MyChart imports merge cleanly.")
                .font(.caption2).foregroundStyle(.secondary)
        }
        .padding(CarePlusSpacing.md)
        .background(CarePlusPalette.surfaceElevated, in: RoundedRectangle(cornerRadius: CarePlusRadius.md))
        .padding(.horizontal, CarePlusSpacing.lg)
        .padding(.vertical, CarePlusSpacing.sm)
    }

    private func bmiLabel(_ b: Double) -> String {
        switch b {
        case ..<18.5:  return "Under"
        case ..<25:    return "Normal"
        case ..<30:    return "Over"
        default:       return "Obese"
        }
    }

    private func bmiColor(_ b: Double) -> Color {
        switch b {
        case ..<18.5:  return CarePlusPalette.info
        case ..<25:    return CarePlusPalette.success
        case ..<30:    return CarePlusPalette.warning
        default:       return CarePlusPalette.danger
        }
    }

    // MARK: - Connected sources row

    private enum SourceStatus { case connected, notConnected, add }

    private func connectedRow(symbol: String, label: String,
                              status: SourceStatus) -> some View {
        HStack {
            Image(systemName: symbol).foregroundStyle(CarePlusPalette.careBlue).frame(width: 24)
            Text(label)
            Spacer()
            switch status {
            case .connected:
                Image(systemName: "checkmark.circle.fill")
                    .foregroundStyle(CarePlusPalette.success)
            case .notConnected:
                Text("Connect").font(.caption.weight(.semibold))
                    .foregroundStyle(CarePlusPalette.careBlue)
            case .add:
                Text("Add").font(.caption.weight(.semibold))
                    .foregroundStyle(CarePlusPalette.careBlue)
            }
        }
    }
}
